server/src/main/java/org/opensearch/index/store/IndexStoreListener.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.common.annotation.PublicApi;
+import org.opensearch.core.index.Index;
+import org.opensearch.core.index.shard.ShardId;
+import org.opensearch.env.NodeEnvironment;
+import org.opensearch.index.IndexSettings;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A listener that is executed on per-index and per-shard store events, like deleting shard path
+ *
+ * @opensearch.api
+ */
+@PublicApi(since = "2.19.0")
+public interface IndexStoreListener {
+    default void beforeShardPathDeleted(ShardId shardId, IndexSettings indexSettings, NodeEnvironment env) {}
+
+    default void beforeIndexPathDeleted(Index index, IndexSettings indexSettings, NodeEnvironment env) {}
+
+    IndexStoreListener EMPTY = new IndexStoreListener() {
+    };
+
+    /**
+     * A Composite listener that multiplexes calls to each of the listeners methods.
+     */
+    final class CompositeIndexStoreListener implements IndexStoreListener {
+        private final List<IndexStoreListener> listeners;
+        private final static Logger logger = LogManager.getLogger(CompositeIndexStoreListener.class);
+
+        public CompositeIndexStoreListener(List<IndexStoreListener> listeners) {
+            this.listeners = Collections.unmodifiableList(listeners);
+        }
+
+        @Override
+        public void beforeShardPathDeleted(ShardId shardId, IndexSettings indexSettings, NodeEnvironment env) {
+            for (IndexStoreListener listener : listeners) {
+                try {
+                    listener.beforeShardPathDeleted(shardId, indexSettings, env);
+                } catch (Exception e) {
+                    logger.warn(() -> new ParameterizedMessage("beforeShardPathDeleted listener [{}] failed", listener), e);
```

Comment:
It seems like potentially a big change to swallow exceptions. The alternatives are:
- let the first exception propagate and not execute subsequent listeners
- attempt all listeners and accumulate exceptions, throwing one with all other exceptions added as supressed

What do you think?

Reply:
That's a good point. So referencing `SearchOperationListener` is seems like some of the handlers swallow the exception:
https://github.com/opensearch-project/OpenSearch/blob/978d14e867c41b224303b2e77745495336568f79/server/src/main/java/org/opensearch/index/shard/SearchOperationListener.java#L313-L322

While some of the handlers accumulate them:
https://github.com/opensearch-project/OpenSearch/blob/978d14e867c41b224303b2e77745495336568f79/server/src/main/java/org/opensearch/index/shard/SearchOperationListener.java#L335-L346

All of the pre/post hooks seem to do swallowing (which IMO is similar to the `IndexStoreListener` hooks), while the one that doesn't swallow is a validation hook, so in that case it makes sense to not swallow a validation failure.

Reply:
So in the existing `IndexStoreListener` implementation `FileCacheCleaner` we are already swallowing any failure exceptions in the hooks, so I guess the remaining question is if the composite listener should swallow exceptions, or if it should be left up to the individual implementation to make that decisions. The convention in the existing listeners (`SearchOperationListener`, `IndexingOperationListener`) seems to be swallow with a warning so I am leaning in that direction.

Reply:
I am also inclining towards we should not throw the exception as the listener here is to let implementation know about the stage the code flow is in, instead of providing a way to break that flow (on some condition) which is what validation sort of callbacks are providing. That is also sort of suggested by the callback methods signature which doesn't suggest throwing any checked exceptions.